### ISSUE TITLE
Done: Add a ruler

### ISSUE THREAD

**Issue by MBPhys** [napari/napari#4200]
## 🚀 Feature
Add a ruler to measure distance in world space. 

## Motivation
There were a lot of requests to implement this and I said that I can do this in the community meeting. For me, this is also useful in napari. Now, there is a working implementation :smiley: and I want to share my design suggestion.
@tlambert03, @sofroniewn 

## Pitch
The ruler lives on the scale bar overlay. I use the interaction box overlay and transform the scale bar in the world space in order to measure the distances there. Now, the interaction box is no longer only correlated with the active layer, but can also be correlated with overlay objects like the scale bar, how I did it. The ruler can be activated when the scale bar is activated and a key is pressed.
See the attached video.

https://user-images.githubusercontent.com/83079938/156732507-bfc39a2c-e281-4285-813f-7ce7e76aa1b6.mp4

**Comment by psobolewskiPhD** [comment]
Love it!
❤️

**Comment by sofroniewn** [comment]
Wow this is so cool @MBPhys!! I like how you're re-using the scale bar. I wonder though would people ever like multiple rulers? Or would people ever like to have both a ruler and a scale bar? I'm not sure. I think this is a fine place to start, but very curious what others think, cc @napari/core-devs? I think a PR would be great though - this will be very useful for a lot of folks!!

**Comment by brisvag** [comment]
I agree that a ruler would be amazing, but I don't think resizing and rotating the scalebar is a good approach. I'm sure what everyone really wants is to be able to click and drag from x to y and get the distance, without having to fight with handles and stuff. I would instead suggest to reuse the measurement logic from the scale bar, but recreate a new overlay.

Also, a reminder that there were attempts like #3763 to unify overlays as things that live in canvas space, and a changes like this are going against that. That's not a problem per se (happy to drop that idea if this makes more sense!), I just wanted to connect the discussions :)

Another thing worth thiking about is a ruler in 3D that lives in scene space (cc @kevinyamauchi @alisterburt).

**Comment by sofroniewn** [comment]
>  I'm sure what everyone really wants is to be able to click and drag from x to y and get the distance, without having to fight with handles and stuff. I would instead suggest to reuse the measurement logic from the scale bar, but recreate a new overlay.

Good point, click and drag would be good to have too - maybe nice to also allow resizing afterwards, so maybe best to have the ruler be its own overlay and have both click and drag and an interaction box. Regardless, it would be nice to pick up #3763 again and get that merge

**Comment by psobolewskiPhD** [comment]
If I had my druthers this would be part of Shapes layer tools--more intuitive for me anyways. A ruler icon as an option on the line? Could indicate with a corner like Adobe tools.

**Comment by alisterburt** [comment]
wow! super cool @MBPhys :)

I think overall I agree with @brisvag, I would prefer if this was a separate entity from the scale bar - with click and drag to define then with a handle on each end for manipulation. 

Starting with 2D is great and this will allow us to add 3D later. @brisvag is right that we don't have a obvious place to put 'things which live in scene space' right now. @psobolewskiPhD I like where you're coming from - a potential intermediate solution could be to add a context menu when we right click on a line which shows the length of the line in both data and world coordinates (with units from scale bar if set), would be curious if others like/dislike this idea!

**Comment by psobolewskiPhD** [comment]
> a potential intermediate solution could be to add a context menu when we right click on a line which shows the length of the line in both data and world coordinates (with units from scale bar if set), would be curious if others like/dislike this idea!

Anything that gets use closer!
I think there are things a user (like me) can want: 
1) to be able to quickly get the size of an object they are viewing. I confess I do this a lot, usually by zooming and repositioning so the scale bar is over my object of interest. Then I eye ball it. 🤣 
2) to be able to indicate the size/distance visually on the image with a text label (optional). This is related to the above. One could also tweak the line ends to make them more ruler like, etc.
3) to be able to repeatedly `measure` the length/diameter of objects. Draw -> measure -> Draw -> measure, so on. There's an example that's a skeleton for that and this may be better for a plugin/widget.

**Comment by MBPhys** [comment]
Thanks for your fast feedback :smiley: 
I agree with you, @brisvag, to add the option of drawing a ruler from point x to point y, additionally and create an independent overlay object for the ruler.
Furthermore, I think unifying the overlays is a good idea, too. This looks like a lot of work:) If you want I can help, @brisvag? How I understand the overlays, there are objects they live in world space like the interaction box or canvas space like the scale bar. But I think a ruler or an interaction box makes more sense in world space for me because they interact with them, or? Maybe a good approach could be to have the choice, but this is already the case,
https://github.com/napari/napari/blob/6ce1968d0924b3b104c3fe3ec7595fd17215a92b/napari/_qt/qt_viewer.py#L386-L414

the parent self.view for canvas space or self.view.scene for world space. There is just a linear transformation between them to compare the systems. If I am wrong, please correct me, @sofroniewn @brisvag? 

@alisterburt @psobolewskiPhD 
This is a good point to make a ruler on the top of a line shape object as additional property for the object. This can be
extended to a rectangle and so on, too. I love the idea @alisterburt, because it is light, too.
Nonetheless, my idea behind this was to have a ruler, which looks like a ruler and is fully independent of the layer interaction, just lives on the viewer in the world space for each 2D perspective and you do not need new buttons and icons. This implementation works for nd layers like the interaction box but is not yet supported for the 3D view like the interaction box. This is the case for drawing objects like lines for shape layers, too. This to add could be the next big step, but I think is not so easy and need a discussion previously that there are not going against and be consistent.
What do you think?

Another idea would be to have measurements like a ruler independent of overlays as an individual measurement layer? From my side, there is the question in which direction napari should go in the future? In the direction of imageJ with a lot of built-in stuff for instance or in the direction of a light fast viewer as it is now? For elementary measurements, I think there are a lot of opportunities like adding, cropping, and so on. This is a question, which the napari community has to answer :smile: In my opinion napari is a beautiful light viewer in python and the very good plugin system provides required stuff on top selecting individually by users so that they are not flooded with icons/buttons/options, which they do not need at all for their usage.

**Comment by sofroniewn** [comment]
> the parent self.view for canvas space or self.view.scene for world space. There is just a linear transformation between them to compare the systems. If I am wrong, please correct me, @sofroniewn @brisvag?

Yes i think that is right

> Nonetheless, my idea behind this was to have a ruler, which looks like a ruler and is fully independent of the layer interaction, just lives on the viewer in the world space for each 2D perspective and you do not need new buttons and icons. 

I think it is nice to have the separate ruler as people shouldn't be forced to add a shapes layer just to measure something - but we might also want to make is super easy to add a text box to each line with its length and a text box to each polygon with its area - that would be a fun feature

> In my opinion napari is a beautiful light viewer in python and the very good plugin system provides required stuff on top selecting individually by users so that they are not flooded with icons/buttons/options, which they do not need at all for their usage.

yeah - i do like that simplicity too. i'm curious what others think here too

**Comment by jinxsfe** [comment]
Hi, I have a question about how you can resize the scaler size; how do you employ an interaction box? thanks.

**Comment by brisvag** [comment]
@jinxsfe this is a really old issue, and the "interaction box" no longer exists in that form, I don't think you could make this work.

Actually, this reminds me that this ruler tool is a great candidate for the new overlays system, and should be relatively easy to implement...
